### PR TITLE
Bug fix and improvement to aggregate reports pipeline

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Fixes bug in aggragate reports pipeline due to incomolet artifacts downloaded from net - core pipeline.
- Improvements to aggregate reports pipeline.

### CODE REVIEW

File: eng/ignore-links.txt [hunk 1]
```diff
@@ -5,3 +5,4 @@ https://github.com/Azure/azure-sdk-for-net-pr/tree/feature/IoT-ADT/sdk/digitaltw
 https://github.com/Azure/azure-sdk-for-net-pr/blob/feature/IoT-ADT/sdk/iot/Azure.Iot.DigitalTwins/samples/DigitalTwinServiceClientSample/DTDL/Relationships/HospitalRelationships.json
 https://www.nuget.org/packages/Azure.ResourceManager.DigitalTwins
 https://contoso.azureedge.net/urlsigning/test?expires=2145916800&amp;keyid:key1&amp;signature=iTsrLX9rVAIJkSahBA_j5o9Azf5-j331ohxDR1Gx2js=
+https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers
```
Comment (weshaggard) [inline eng/ignore-links.txt:8]: Why is this added?

Comment (chidozieononiwu) [reply eng/ignore-links.txt:8]: That is a false positive that was flagged by the link checker. I added it here just to be able to test the PR.

Comment (weshaggard) [reply eng/ignore-links.txt:8]: I think it was flagged because of "en-us" being in the link which is actually an issue we should fix.

File: eng/pipelines/aggregate-reports.yml [hunk 2]
```diff
@@ -31,61 +22,61 @@ stages:
...
+              /p:IncludeTests=false
+              /p:PublicSign=false
+              /p:OfficialBuildId=$(OfficialBuildId)
+              /p:SkipDevBuildNumber=true
+              /p:Configuration=$(BuildConfiguration)
+              /p:CommitSHA=$(Build.SourceVersion)
+              /p:ArtifactsPackagesDir=$(Build.ArtifactStagingDirectory)
+            displayName: "Build and Package"
+
+          #- pwsh: |
```
Comment (weshaggard) [inline eng/pipelines/aggregate-reports.yml:40]: I assume this is commented out because you are testing.

File: eng/ignore-links.txt [hunk 3]
```diff
@@ -4,4 +4,4 @@ https://github.com/Azure/azure-digital-twins/blob/private-preview/Documentation/
 https://github.com/Azure/azure-sdk-for-net-pr/tree/feature/IoT-ADT/sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/DTDL/DigitalTwins
 https://github.com/Azure/azure-sdk-for-net-pr/blob/feature/IoT-ADT/sdk/iot/Azure.Iot.DigitalTwins/samples/DigitalTwinServiceClientSample/DTDL/Relationships/HospitalRelationships.json
 https://www.nuget.org/packages/Azure.ResourceManager.DigitalTwins
-https://contoso.azureedge.net/urlsigning/test?expires=2145916800&amp;keyid:key1&amp;signature=iTsrLX9rVAIJkSahBA_j5o9Azf5-j331ohxDR1Gx2js=
```
Comment (weshaggard) [inline eng/ignore-links.txt:7]: Let's undo the changes to this file now.

File: eng/scripts/DownloadDevPackages.ps1 [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    [Parameter(Mandatory=$True)]
+    [string] $NupkgDestination,
+    [string] $NugetSource="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json",
+    [string] $FallBackSource = "https://api.nuget.org/v3/index.json"
+)
+
+. (Join-Path $PSScriptRoot ".." common scripts common.ps1)
+
+$allPackages = Get-AllPkgProperties
+$trackTwoPackages = $allPackages.Where{ $_.SdkType -eq "client" }
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:13]:
```suggestion
$trackTwoPackages = $allPackages.Where{ $_.IsNewSdk }
```
I think track 2 mgmt is also intersting.

File: eng/scripts/DownloadDevPackages.ps1 [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  -Prerelease `
+  -Source $NugetSource `
+  -FallBackSource $FallBackSource `
+  -DependencyVersion Ignore `
+  -DirectDownload `
+  -NoCache `
+  -ExcludeVersion
+}
+
+$nupkgFilesPath = Join-Path $WorkingDirectory $NupkgDestination
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:32]: Have you confirmed that this only puts the packages we are interested in in the path?

File: eng/scripts/DownloadDevPackages.ps1 [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+{
+  nuget install $package.Name `
+  -OutputDirectory $nugetPackagesPath `
+  -Prerelease `
+  -Source $NugetSource `
+  -FallBackSource $FallBackSource `
+  -DependencyVersion Ignore `
+  -DirectDownload `
+  -NoCache `
+  -ExcludeVersion
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:29]: Do we really wish to exclude the version?

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1:29]: I think that just excluded the version from file/directory name.

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1:29]: I guess it adds no value, So I removed it.

File: eng/scripts/DownloadDevPackages.ps1 (hunk 5 above)
Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1:32]: Yes, Only packages marked as IsNewSdk are downloaded without their dependencies  and this just copies the nupkg file out to where it can be used for other processes.

**Comment by weshaggard** [comment]
@chidozieononiwu looks like you are still doing some testing here so please ping me when you are ready for me to review.

File: eng/ignore-links.txt (hunk 3 above)
Comment (weshaggard) [reply eng/ignore-links.txt:7]: Looks like there is still some edit here where remove the eof marker.

File: eng/pipelines/aggregate-reports.yml [hunk 7]
```diff
@@ -122,15 +105,27 @@ stages:
...
+              sourcePath: '$(Build.ArtifactStagingDirectory)/reports/*'
               azureSubscription: 'Azure SDK Artifacts'
               destination: AzureBlob
               storage: azuresdkartifacts
               containerName: 'azure-sdk-for-net'
               blobPrefix: dependencies
+              AdditionalArgumentsForBlobCopy: '--exclude-pattern=*data.js*'
+
+          - task: AzureFileCopy@4
+            displayName: 'Upload dependency report (2)'
```
Comment (weshaggard) [inline eng/pipelines/aggregate-reports.yml:120]: Why do we have 2 different uploads?

Comment (chidozieononiwu) [reply eng/pipelines/aggregate-reports.yml:120]: its because am making copies to different directories int he blob container. I couldn't figure out how to use File Copy to do it in ione step.

File: eng/ignore-links.txt (hunk 3 above)
Comment (chidozieononiwu) [reply eng/ignore-links.txt]: I have fixed that. Thanks

File: eng/pipelines/aggregate-reports.yml (hunk 7 above)
Comment (weshaggard) [reply eng/pipelines/aggregate-reports.yml:120]: We should give the names more meaning beyond 1 and 2 and tell what they are for. For example it looks like one is reports and the other is graphs.

File: eng/scripts/DownloadDevPackages.ps1 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    [Parameter(Mandatory=$True)]
+    [string] $NupkgDestination,
+    [string] $NugetSource="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json",
+    [string] $FallBackSource = "https://api.nuget.org/v3/index.json"
+)
+
+. (Join-Path $PSScriptRoot ".." common scripts common.ps1)
+
+$allPackages = Get-AllPkgProperties
+$trackTwoPackages = $allPackages.Where{ $_.IsNewSdk }
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:13]:
```suggestion
$trackTwoPackages = $allPackages.Where({ $_.IsNewSdk })
```
I believe this needs parens, did it work without them?

**Review by weshaggard — DISMISSED** [review]
Left a couple comments but otherwise looks good. I assume the aggregate-reports pipeline is now functioning after your changes.

File: eng/scripts/DownloadDevPackages.ps1 [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+Push-Location $WorkingDirectory
+$nugetPackagesPath = Join-Path $WorkingDirectory nugetPackages
+New-Item -Path $WorkingDirectory -Type "directory" -Name "nugetPackages"
+
+$packagesInFeed = Find-Package -Source $NugetSource -AllVersion -AllowPrereleaseVersions
+
+foreach ($package in $trackTwoPackages)
+{
+  $allVersionsInFeed = $packagesInFeed.Where({ ($_.Name -eq $package.Name) })
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:22]: May be worth adding this in a helper in language-settings so we can share it across here and smoke-tests.

FYI @ckairen @benbp

File: eng/scripts/DownloadDevPackages.ps1 [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  # Use the latest alpha version where it is available
+  if ($alphaVersionsInFeed.Count -gt 0)
+  {
+    $alphaVersionsSorted = [AzureEngSemanticVersion]::SortVersionStrings($alphaVersionsInFeed.Version)
+    $latestDevVersion = $alphaVersionsSorted[0]
+  }
+  else
+  {
+    $allVersionsSorted = [AzureEngSemanticVersion]::SortVersionStrings($allVersionsInFeed.Version)
+    $latestDevVersion = $allVersionsSorted[0]
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:34]: We need to verify that allVersionsInFeed is not empty. I think this maybe the error you are hitting currently.

File: eng/scripts/DownloadDevPackages.ps1 [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+. (Join-Path $PSScriptRoot ".." common scripts common.ps1)
+
+$allPackages = Get-AllPkgProperties
+$trackTwoPackages = $allPackages.Where({ $_.IsNewSdk })
+
+Push-Location $WorkingDirectory
+$nugetPackagesPath = Join-Path $WorkingDirectory nugetPackages
+New-Item -Path $WorkingDirectory -Type "directory" -Name "nugetPackages"
+
+$packagesInFeed = Find-Package -Source $NugetSource -AllVersion -AllowPrereleaseVersions
```
Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:18]: We should see how slow this is once we are done as it might end up taking a while.

Comment (weshaggard) [inline eng/scripts/DownloadDevPackages.ps1:18]: After a little local testing I think you can simply do the following:

```
$packageVersion = [AzureEngSemanticVersion]::ParseVersionString($package.Version)
$packageVersion.IsPreReleaseLabel = $false # clear prerelease so ToString strips it prerelease label.  
$packageVersionNoPrerelease $packageVersion.ToString()

For some reason I couldn't get this to work locally with the v3 feed but the v2 one worked correctly
#$NugetSource="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json"
$NugetSource="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v2"

Install-Package -Name $package.Name -Source $NugetSource -AllowPrereleaseVersions -MinimumVersion "${packageVersionNoPrerelease}-alpha" -MaximumVersion "${packageVersionNoPrerelease }-alphb" -Destination "$nugetPackagesPath" -Force -SkipDependencies
```

This essentially will pull the latest alpha version for the package version that is checked into the repo. If there is no current new version yet it will fail.

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1]: This was slow and it did not return all the packages. I sitche to calling the DevOps Feed api.

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1]: THis still requires using wildcards for the version which does not work.

Comment (weshaggard) [reply eng/scripts/DownloadDevPackages.ps1]: It is not using wildcards it is using a version range and it does work as I tested it locally. Can you help me understand why it doesn't?

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1]: This method is still very unreliable. I does not get all the packages. It give an error `no match for package` for some packages wve though they are clearly present in the feed. There is also the issue of where it matches more that one version which cause it to fail with `Unable to install, multiple packages matched`

Comment (weshaggard) [reply eng/scripts/DownloadDevPackages.ps1]:
```
# Install-Package requires a v2 nuget feed end-point
$NugetSource="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v2"
...
  $packageVersion = [AzureEngSemanticVersion]::ParseVersionString($package.Version)
  $packageVersion.IsPreRelease = $false # clear prerelease so ToString strips it prerelease label.
  $packageVersionNoPrerelease = $packageVersion.ToString()

  # To workaround some older invalid packages (i.e. Iot->IoT renamed packages) start the alpha version range to start at 6/21
  Write-Verbose "Installing alpha version for $($package.Name) ${packageVersionNoPrerelease}"
  $installedPackage = Install-Package -Name $package.Name `
    -Source $NugetSource `
    -MinimumVersion "${packageVersionNoPrerelease}-alpha.202106" `
    -MaximumVersion "${packageVersionNoPrerelease}-alphb" `
    -Destination "$nugetPackagesPath" `
    -AllowPrereleaseVersions `
    -Force `
    -SkipDependencies `
    -ErrorAction Ignore

  if ($installedPackage) {
    Write-Host "Installed $($package.Name) $($installedPackages.Version)"
  }
  else {
    Write-Warning "Did not find any matching package alpha versions for $($package.Name) ${packageVersionNoPrerelease}"
  }
```

I tested that code locally and it works as expected without issues.

Comment (chidozieononiwu) [reply eng/scripts/DownloadDevPackages.ps1]: Thanks. this works.

File: eng/scripts/dependencies/generate-deps.ps1 [hunk 12]
```diff
@@ -7,6 +7,11 @@ param(
 mkdir $ProjectRefPath -force | out-null
 $projRefsPath = Join-Path $ProjectRefPath "azure-sdk-projectreferences.props"
 
+$ignoreList = @(
+  "Azure.Monitor.OpenTelemetry.Exporter",
```
Comment (weshaggard) [inline eng/scripts/dependencies/generate-deps.ps1:11]: Please file issues for these and leave a comment here with a link to the issues.

**Review by weshaggard — APPROVED** [review]
Changes look good but lets make sure we have a green aggregate-reports pipeline before merging.
